File: src/x/crazyjvm/decorate/j/Decorate.java
package x.crazyjvm.decorate.j;

/**
 * Created by chenchao on 15/10/12.
 */
public class Decorate {
}

class FileOutputStream implements OutputStream {
    public void write(byte b){}
    public void write(byte[] b){}
}

abstract class OutputStreamDecorator implements OutputStream {
    protected final OutputStream delegate;

    protected OutputStreamDecorator(OutputStream delegate) {
        this.delegate = delegate;
    }

    public void write(byte b) { delegate.write(b); }
    public void write(byte[] b) { delegate.write(b); }
}

class BufferedOutputStream extends OutputStreamDecorator {
    public BufferedOutputStream(OutputStream delegate) {
        super(delegate);
    }

    public void write(byte b) {
        // ...
        //delegate.write(buffer)
    }
}

//new BufferedOutputStream(new FileOutputStream("foo.txt"))

File: src/x/crazyjvm/decorate/s/Decorate.scala
package x.crazyjvm.decorate.s

/**
 * Created by chenchao on 15/10/12.
 */
class Decorate {

}


trait OutputStream {
  def write(b: Byte)
  def write(b: Array[Byte])
}

class FileOutputStream(path: String) extends OutputStream {
  def write(b: Byte){}
  def write(b: Array[Byte]){}
}

trait Buffering extends OutputStream {
  abstract override def write(b: Byte) {
    // ...
    //super.write(buffer)
  }
}

//new FileOutputStream("foo.txt") with Buffering // with Filtering, ...